### ISSUE TITLE
Learning appliance models from multiple homes

### ISSUE THREAD

**Issue by nipunreddevil** [nilmtk/nilmtk#463]
Our current supervised algorithm implementations- CO and FHMM learn the model from the same home on which we intend to disaggregate. This isn't the most realistic scenario. 

I'll start with Oli's excellent post on ["What even is supervised/unsupervised disaggregation?"](http://blog.oliverparson.co.uk/2015/05/what-even-is-supervisedunsupervised.html) as a reference point. What we want to do is to choose a train set and learn the "general" appliance model across these homes.

Let's us assume that we know the number and type of appliances in the test homes.

We can use the following procedure:

1. Divide data set into training and test based on homes. Say, out of 50 homes, we train on 25 homes and test on remaining 25.
2. Say in 20 out of 25 homes we have a fridge. We now feed these 20 traces to a HMM to learn the fridge model. Similarly, we do so for all other appliances.
3. For testing- for each home in test set, we create a FHMM corresponding to the HMMs for the appliances present in the home.
4. For appliances having a single instance, the evaluation should be straightforward. For appliances with multiple instances, we'll need to do a matching with ground truth.

How does this sound @JackKelly, @oliparson  ?

This approach seems to be different from the Disaggregator interface that we've specified thus far! Any thoughts?

**Comment by oliparson** [comment]
This sounds great Nipun. However, remember that we're probably going to have quite a few different scenarios for training/testing:
1.  Supervised - appliance level data available from the house in which disaggregation is performed
2.  Blind with appliance list - multiple examples of appliance level data available from houses other than the one in which disaggregation is performed, and a list of target appliances are supplied for the test house. Furthermore, aggregate data should be available from the test house.
3.  Completely blind - multiple examples of appliance level data available from houses other than the one in which disaggregation is performed, but nothing is known about the target appliances present in the test house. Furthermore, aggregate data should be available from the test house. Disaggregated appliances could be either labelled or unlabelled.

As you might have noticed, I'm starting to favour the term 'blind' over the term 'unsupervised' for the reason that blind disaggregation can use appliance-level data from homes other than the one in which we want to do disaggregation!

We currently do a good job of supporting approach 1, but approach 2 is what you're proposing here. Although approach 3 is the most realistic, let's start by supporting approach 2 first :)

I think our Disaggregator interface isn't too far away from supporting approach 2, since a meter group can support an arbitrarily large collection of appliance instances. However, we need to be careful to specify whether the site-meter or appliance meters come from the test house or non-test houses!

**Comment by JackKelly** [comment]
Totally agree that training on multiple homes is important.  I did this in my Neural NILM work (with both synthetic and real aggregate data)

Also I very much like the term 'blind' as it side-steps the confusion around 'unsupervised'.

My hunch  is that perhaps we don't need to change the `Disaggregator` API much.  We can already call `train_on_chunk()` to iteratively train a `Disaggregator`.  So perhaps all we need is some simple pre-processing tools to split up a `DataSet` into test houses and train houses and feed each train house one-by-one to a `Disaggregator` using `train_on_chunk()`.

Then, for aligning the `Disaggregator`'s output with ground truth for metrics, I think this alignment should be the metric's responsibility.

(Haven't given this a huge amount of thought... my mind is very much on my PhD thesis write-up at the moment!)

**Comment by JackKelly** [comment]
one further thought: there are a large number of Disaggregation scenarios.  I wonder if, instead of trying to support every scenario out-of-the-box, instead we should think more along the lines of making simple building blocks which users can assemble to help them with whatever scenario they have in mind.  Not entirely sure how to do this though!

**Comment by nipunreddevil** [comment]
Thanks for the feedback guys! I'll make a start on this tomorrow as I need it for my research. I'll start with a hacky solution based just on Pandas data frames and get back to nilmtk later.

**Comment by JackKelly** [comment]
> I'll start with a hacky solution based just on Pandas data frames and get back to nilmtk later

yeah, I think that's a good approach :)

**Comment by oliparson** [comment]
Related: #195

**Comment by nipunbatra** [comment]
Hi @oliparson , @JackKelly, 
I have an initial implementation of FHMM training and disaggregation across multiple homes. The code is a bit scratchy. There is a lot of code duplication as well. But, my main aim was to demo the functionality.
Here is the IPython notebook, which I'll clean up and document in a while: https://github.com/nipunbatra/nipunbatra.github.io/blob/master/downloads/notebooks/nilmtk_hmm_across_homes.ipynb

Please let me know your thoughts. At the moment, I have kept `train` and `train_across_homes` as two separate methods of the FHMM class. Depending on the use case, one will call the pertinent method. Similar is the case with `disaggregate` and `disaggregate_across_homes`.

`train_across_homes` currently has the following arguments:
1. list of homes
2. list of appliances for which we want to train the model
3. activation threshold for choosing homes for training (we may want homes with atleast some dense usage of the appliance)
4. dataset

On similar lines, `disaggregate_across_homes` will take as arguments:
1. dataset
2. list of homes

One important thing to note. Our learnt model may have appliances for which we don't have ground truth data. I am not writing the data from those appliances in disaggregated data store. It was causing issues further down the pipeline.

I've pushed the changes to a new branch. Changes visible here: https://github.com/nilmtk/nilmtk/compare/fhmm_multiple_homes


@JackKelly : I really hacked my way around getting the code to work :) How do you think we can reduce the redundancy between the `disaggregate` and the proposed `disaggregate_across_homes` method. I think the trouble comes in the metadata handling portion. 

After a round of code reviews, we may merge this branch on the main branch. Then, getting CO to work across homes should be simple as well.

**Comment by JackKelly** [comment]
Hi,

Good work :)

I'm afraid I'm not sure I'll have a chance to look at the code any time soon as I'm really behind in my thesis writing.

A few quick thoughts:

1) I think we should be consistent in our use of terminology when it comes to 'homes' versus 'buildings'.  In most of NILMTK we use 'buildings' instead of 'homes'.
2) Rather than add another method name that users have to learn, I wonder if we could just modify `disaggregate` and `train` so that if they receive a `DataSet` object (instead of a `MeterGroup` object) then the `disaggregate` or `train` will operate over then entire dataset.  This should need much change in the code: just add a few lines to the start of `disaggregate` and `train` to check in the incoming object is a `DataSet` or `MeterGroup` object.   What do you think?  I'm acutely aware that NILMTK already has a very large, sprawling API and it might help users if we kept the API as simple as possible.  Sticking with the same method names also makes it easier to do some experiments on a single house and then easily switch to running disaggregation or training over an entire dataset.  What do you think?

**Comment by nipunbatra** [comment]
Thanks for the inputs @JackKelly 

Both of these make sense. I think I originally wanted to use a single method, but, I think method overloading isn't really a thing in Python?

**Comment by JackKelly** [comment]
>  I think method overloading isn't really a thing in Python?

It's true that Python doesn't natively support it.  But there are work arounds...

I usually do this sort of thing (not entirely elegant but it works!):

```python
def add(self, other):
    if isinstance(other, Foo):
        ...
    elif isinstance(other, Bar):
        ...
    else:
        raise NotImplementedError()
```

but I've just found that there might be a more elegant way of doing it:

http://stackoverflow.com/a/29091980

**Comment by nipunbatra** [comment]
Thanks @JackKelly 
Multi dispatch looks interesting! I'll give it a shot tomorrow.

**Comment by nipunbatra** [comment]
I did try multiple dispatch. It, however, doesn't like *args and **kwargs and wants a fixed function definition. So, it doesn't help our cause.

I also tried using the more naive `isinstance` based solution. Something like:

```python
def train(self, meter_ds, *args, **kwargs):
        if isinstance(meter_ds, nilmtk.DataSet):
            self.train_across(meter_ds, *args, **kwargs)
        elif isinstance(meter_ds, nilmtk.MeterGroup):
            self.train_single(meter_ds, *args, **kwargs)
```

I again faced issues due to `*args` and `*kwargs`, such as

```python
g= FHMM()
g.train(ds, list_of_homes=range(1, 5), list_of_appliances=['air conditioner', 'fridge'],sample_period=600)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-d0299d573ee8> in <module>()
----> 1 g.train(ds, list_of_homes=range(1, 5), list_of_appliances=['air conditioner', 'fridge'],sample_period=600)

/Users/nipunbatra/git/nilmtk/nilmtk/disaggregate/fhmm_exact.py in train(self, meter_ds, *args, **kwargs)
    190     def train(self, meter_ds, *args, **kwargs):
    191         if isinstance(meter_ds, nilmtk.DataSet):
--> 192             self.train_across(meter_ds, *args, **kwargs)
    193         elif isinstance(meter_ds, nilmtk.MeterGroup):
    194             self.train_single(meter_ds, *args, **kwargs)

TypeError: train_across() takes at least 4 arguments (3 given)
```

So, I don't know a clean way of handling method overloading when we don't have fixed signature. 

Aside: I was also thinking if we can be more explicit and let the user know if they're training across or on single home. This will break previous versions, but, there will be less magic under the hood and the user will clearly know what they're doing.

**Comment by JackKelly** [comment]
Yeah, there are definitely advantages to being more explicit.  Although I'd personally still favour having a single method name for training across multiple buildings or training across a single building (so the same user code can be used for both cases, and so the user sees a simpler API).  But if there's no clean way to do it then just do what ever you think's best.  I must admit that my head is in 'PhD writing mode'  at the moment and I'm struggling to think through all the NILMTK implications :)  So just do whatever you think's best.  Although I'd also favour *not* breaking backwards compatibility unless we really really have to.  So I'd suggest we keep a `train` method for training on a single building and a `train_across_buildings` method for training on multiple buildings.

We could add a `train_on_single_building` method and then modify `train` so it raises a deprecated warning and then just calls `train_on_single_building`.  But that feels like more pain than it's worth!  i have to say that my tastes are changing a bit from having verbose, large APIs to having compact APIs.  But that might just be personal taste.

**Comment by nipunbatra** [comment]
Okay. For now, I'll leave the older method unaltered. I'll add the new method `train_across_buildings`. I think the main challenge I'm facing is more of a technical challenge. I did try using a single train method. But, maybe, today is a bad day for nilmtk implementation :) 

I'll merge the train and disaggregate across multiple buildings code into the main branch. I'll probably wait till we're able to find some solution before writing a blog post on the new feature.

**Comment by JackKelly** [comment]
>  I did try using a single train method. But, maybe, today is a bad day for nilmtk implementation

Yeah, I'm pretty certain it is possible to do what you want by using a single method and then using `if isinstance()` but I can't think of exactly how to get round the problem that you mentioned earlier today :)  I'm very tempted to stop work on thesis and do some Python tinkering but that's probably not an idea that my supervisor would like ;)

**Comment by nipunbatra** [comment]
@JackKelly : Ha ha :) I think it is okay to keep the method as it is for now. Those who want to really use it, won't mind changing to a new API (when we do so in the future).

**Comment by beautifwhale** [comment]
Hi @nipunbatra this is an old issue but I am trying to use the FHMM algorithm on the REDD dataset across all buildings 1-5. I have noticed the `fhmm.train_across_buildings` function in the source, and I'm aware that this feature has not been released yet. But I'm wondering is there an example of this working that I could review. For example, train the FHMM model across all appliances in buildings 1-5 in the REDD dataset and then desegregate building 6 mains?
